### ISSUE TITLE
[2.4.5] Incorrect carrier/method code on selected shipping method in GraphQL

### ISSUE THREAD

**Issue by DuckThom** [magento/magento2#36606]
### Preconditions and environment

- Magento 2.4.5
- GraphQL modules are enabled

### Steps to reproduce

1. Set a shipping address on the cart which has an available shipping method with an underscore in the carrier (ie.   sendcloud_checkout)
2. Set the shipping method on the cart
3. Compare the carrier and method codes in `available_shipping_methods` and `selected_shipping_method` on the `cart.shipping_addresses[0]` object returned by GraphQL

GraphQL query:

```graphql
query {
    cart(cart_id: "<cart_id>") {
        shipping_addresses {
            selected_shipping_method {
                carrier_code
                method_code
            }
            available_shipping_methods {
                carrier_code
                method_code
            }
        }
    }
}
```

### Expected result

The returned `selected_shipping_method` matches with one of the `available_shipping_methods` carrier and method codes.

Result based on above query:

```json
{
    "data": {
        "cart": {
            "shipping_addresses": [
                {
                    "selected_shipping_method": {
                        "carrier_code": "sendcloud_checkout",
                        "method_code": "ddc961a8-1824-40ab-88a9-392f2f58e3b5"
                    },
                    "available_shipping_methods": [
                        {
                            "carrier_code": "sendcloud_checkout",
                            "method_code": "ddc961a8-1824-40ab-88a9-392f2f58e3b5"
                        }
                    ]
                }
            ]
        }
    }
}
```

### Actual result

There is a mismatch between the carrier/method codes in the `selectedShippingMethod` and `availableShippingMethods`.

Result based on above query:

```json
{
    "data": {
        "cart": {
            "shipping_addresses": [
                {
                    "selected_shipping_method": {
                        "carrier_code": "sendcloud",
                        "method_code": "checkout_ddc961a8-1824-40ab-88a9-392f2f58e3b5"
                    },
                    "available_shipping_methods": [
                        {
                            "carrier_code": "sendcloud_checkout",
                            "method_code": "ddc961a8-1824-40ab-88a9-392f2f58e3b5"
                        }
                    ]
                }
            ]
        }
    }
}
```

### Additional information

The shipping method is saved on the quote as `method.carrierCode + '_' + method.methodCode`

In the resolver for the selected_shipping_method field, the following line is present:

https://github.com/magento/magento2/blob/ea0cf6395753078e45af8c4a5ddc1ea2c4d91092/app/code/Magento/QuoteGraphQl/Model/Resolver/ShippingAddress/SelectedShippingMethod.php#L53

This will split on the first underscore, which is not correct.

## Possible solution:

```patch
diff --git a/vendor/magento/module-quote-graph-ql/Model/Resolver/ShippingAddress/SelectedShippingMethod.php b/vendor/magento/module-quote-graph-ql/Model/Resolver/ShippingAddress/SelectedShippingMethod.php
--- a/vendor/magento/module-quote-graph-ql/Model/Resolver/ShippingAddress/SelectedShippingMethod.php	
+++ b/vendor/magento/module-quote-graph-ql/Model/Resolver/ShippingAddress/SelectedShippingMethod.php
@@ -38,11 +38,11 @@
             return null;
         }
 
-        list($carrierCode, $methodCode) = explode('_', $address->getShippingMethod(), 2);
-
         /** @var Rate $rate */
         foreach ($rates as $rate) {
             if ($rate->getCode() == $address->getShippingMethod()) {
+                $carrierCode = $rate->getCarrier();
+                $methodCode = $rate->getMethod();
                 $carrierTitle = $rate->getCarrierTitle();
                 $methodTitle = $rate->getMethodTitle();
                 break;

```

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-November** [comment]
Hi @DuckThom ,
Thank you for reporting and collaboration. As per Magento [devdocs](https://developer.adobe.com/commerce/webapi/graphql/schema/cart/mutations/set-shipping-method/) , Graphql supports only below shipping methods by default.
![image](https://user-images.githubusercontent.com/60198231/207858883-18682a6c-31ef-4113-975d-3a4dc8a87fe6.png) and none of the shipping methods have underscore in carrier code to reproduce the issue. Also sendcloud is a 3rd party package and code of the extension is not part of Magento git [repository](https://github.com/magento/magento2). Community cannot able to provide fix for it in this repository. All questions, issue reports and fix for them should be addressed to the corresponding extension owners(support) on the [Magento Market place](https://marketplace.magento.com/extensions.html) page. Kindly recheck the issue on Magento 2.4-develop instance with Graphql supported shipping methods and provide missing steps if the issue is still reproducible.
Thank you.

**Comment by DuckThom** [comment]
I agree that the built-in carriers do not use underscores in the codes, however, it's not mentioned anywhere what the allowed characters are for custom carriers.

IMO, it would be better if it was either mentioned in the dev docs what the allowed characters are or if some validation is added perhaps. As right now, it can lead to cases such as this one where unexpected data is returned.

**Comment by engcom-November** [comment]
Hi @DuckThom ,
Verified the issue again on Magento 2.4-develop instance using custom shipping method using both with / without underscore in carrier code and the issue is reproducible. Hence updating the description and confirming the issue.
**Scenario 1:** Custom shipping method created without underscore in carrier code. No issue. Graphql response retrieved as expected. Method displayed in "available_shipping_methods" and "selected_shipping_method"
![image](https://user-images.githubusercontent.com/60198231/209762421-4def3e28-ae33-48a2-9abc-1b00a0a5e984.png)

**Scenario 2:** Custom shipping method created with underscore in carrier code. Issue: Underscore is missing in Graphql response and the method is not displayed in "available_shipping_methods".
![image](https://user-images.githubusercontent.com/60198231/209762377-5180e4fe-5fcb-4b14-bd3c-91ed29ac45f6.png)

**Comment by Vasudev-22** [comment]
@magento I am working on this

**Comment by engcom-Hotel** [comment]
Hello,

As I can see this issue got fixed in the scope of the internal Jira ticket AC-7550 by the internal team
Related commits: https://github.com/magento/magento2/search?q=AC-7550&type=commits

Based on the Jira ticket, the target version is 2.4.6.

Thanks

**Comment by hostep** [comment]
@engcom-Hotel: either you made a typo, or the info from Jira is incorrect, because those fixes are not included in Magento 2.4.6? It's probably going to be included in 2.4.7(-beta1)?

**Comment by engcom-Hotel** [comment]
Hello @hostep,

That was a typo, I corrected it.

Thanks
